tests/test_pytest_mock.py

```diff
@@ -521,9 +541,67 @@ def __call__(self, x):
     uut.call_like(10)
     spy.assert_called_once_with(10)
     assert spy.spy_return == 20
+    assert spy.spy_return_iter is None
     assert spy.spy_return_list == [20]
 
 
+@pytest.mark.parametrize("iterator", [(i for i in range(3)), iter([0, 1, 2])])
+def test_spy_return_iter(mocker: MockerFixture, iterator: Iterator[int]) -> None:
+    class Foo:
+        def bar(self) -> Iterator[int]:
+            return iterator
+
+    foo = Foo()
+    spy = mocker.spy(foo, "bar")
+    result = list(foo.bar())
+
+    assert result == [0, 1, 2]
+    assert spy.spy_return is not None
+    assert spy.spy_return_iter is not None
+    assert list(spy.spy_return_iter) == result
+    assert spy.spy_return_list == [ANY]
+
+
+@pytest.mark.parametrize("iterable", [(0, 1, 2), [0, 1, 2], range(3)])
+def test_spy_return_iter_ignore_plain_iterable(
+    mocker: MockerFixture, iterable: Iterable[int]
+) -> None:
+    class Foo:
+        def bar(self) -> Iterable[int]:
+            return iterable
+
+    foo = Foo()
+    spy = mocker.spy(foo, "bar")
+    result = foo.bar()
+
+    assert result == iterable
+    assert spy.spy_return == result
+    assert spy.spy_return_iter is None
+    assert spy.spy_return_list == [result]
+
+
+def test_spy_return_iter_unset_in_last_call(mocker: MockerFixture) -> None:
+    class Foo:
+        iterables = [
+            (i for i in range(3)),
+            [3, 4, 5],
+        ]
+
+        def bar(self) -> Iterable[int]:
+            return self.iterables.pop(0)
+
+    foo = Foo()
+    spy = mocker.spy(foo, "bar")
+    result_iterator = list(foo.bar())
+
+    assert result_iterator == [0, 1, 2]
+    assert list(spy.spy_return_iter) == result_iterator
+
+    result_iterable = foo.bar()
+    assert result_iterable == [3, 4, 5]
+    assert spy.spy_return_iter is None
```

Comment:
Minor suggestion to make this clearer:


```suggestion
def test_spy_return_iter_resets(mocker: MockerFixture) -> None:
    class Foo:
        iterables = [
            (i for i in range(3)),
            99,
        ]

        def bar(self) -> Iterable[int]:
            return self.iterables.pop(0)

    foo = Foo()
    spy = mocker.spy(foo, "bar")
    result_iterator = list(foo.bar())

    assert result_iterator == [0, 1, 2]
    assert list(spy.spy_return_iter) == result_iterator

    assert foo.bar() == 99
    assert spy.spy_return_iter is None
```

Mixing iterator and iterables on this test is not necessary for the logic and adds a small bit of confusion.

Reply:
I had a fix ready for this but you were faster